prover/crates/bin/prover_job_monitor/src/job_requeuer/prover_job_requeuer.rs

```diff
@@ -0,0 +1,63 @@
+use std::time::Duration;
+
+use async_trait::async_trait;
+use zksync_periodic_job::PeriodicJob;
+use zksync_prover_dal::{ConnectionPool, Prover, ProverDal};
+
+use crate::metrics::PROVER_JOB_MONITOR_METRICS;
+
+/// `ProverJobRequeuer` is a task that periodically requeues prover jobs that have not made progress in a given unit of time.
+#[derive(Debug)]
+pub struct ProverJobRequeuer {
+    pool: ConnectionPool<Prover>,
+    /// max attempts before giving up on the job
+    max_attempts: u32,
+    /// the amount of time that must have passed before a job is considered to have not made progress
+    processing_timeout: Duration,
+    /// time between each run
+    run_interval_ms: u64,
+}
+
+impl ProverJobRequeuer {
+    pub fn new(
+        pool: ConnectionPool<Prover>,
+        max_attempts: u32,
+        processing_timeout: Duration,
+        run_interval_ms: u64,
+    ) -> Self {
+        Self {
+            pool,
+            max_attempts,
+            processing_timeout,
+            run_interval_ms,
+        }
+    }
+}
+
+#[async_trait]
+impl PeriodicJob for ProverJobRequeuer {
+    const SERVICE_NAME: &'static str = "ProverJobRequeuer";
+
+    async fn run_routine_task(&mut self) -> anyhow::Result<()> {
+        let stuck_jobs = self
+            .pool
+            .connection()
+            .await
+            .unwrap()
```

Comment:
There is quite some boilerplate here. I believe that it can be significantly reduced, if we rework the `PeriodicJob` as follows.

This way you will have a builder, so you have to specify less details in each job impl, you don't have to store pool insdide of each job, and spawning will be much easier.

Since it's a thing specific to this crate, you don't really need to extract it into a separate crate (so that you don't have to refactor existing house keeper code). You can move `PeriodicJob` back to the house keeper so that it dies there, and instead define the structure and crate here. Later, if it will be useful, we will generalize the type over DB type (so that it can be used with `Core` as well) and extract.

```rust
#[async_trait::async_trait]
pub trait DalJob {
    async fn invoke(&self, connection: &mut Connection<Prover>) -> anyhow::Result<()>;
}

// Internal repr of a periodic job
struct PeriodicDalJob {
    job: Box<dyn DalJob>,
    pool: ConnectionPool<Prover>,
    name: String,
    interval: Duration,
}

impl PeriodicDalJob {
    async fn run(&self, stop_receiver: sync::Watch<bool>) -> anyhow::Result<()> {
        while !*stop_receiver.borrow() {
            let mut connection = self.pool.get().await?;
            self.job.invoke(&mut connection).await?;
            tokio::time::sleep(self.interval).await;
            // Do whatever reporing/logging you need
        }
    }
}

pub struct DalJobRunner {
    pool: ConnectionPool<Prover>,
    tasks: Vec<PeriodicDalJob>,
}

impl DalJobRunner {
    pub fn new(pool: ConnectionPool<Prover>) -> Self {
        Self { pool, tasks: Vec::new() }
    }

    pub fn push_job<T: DalJob>(&mut self, name: &str, interval: Duration, job: T) {
        self.tasks.push(PeriodicDalJob { name: name.into(), interval, job: Box::new(job), pool: self.pool.clone() });
    }

    pub fn spawn(self, stop_receiver: tokio::sync::Watch<bool>) -> Vec<tokio::task::JoinHandle<anyhow::Result<()>>> {
        self.tasks.into_iter().map(|task| {
            let pool = self.pool.clone();
            let receiver = stop_receiver.clone();
            tokio::spawn(async move {
                task.run(receiver).await
            })
        }).collect()
    }
}
```